main/src/ca/uwaterloo/flix/language/phase/Inliner1.scala

```diff
@@ -0,0 +1,986 @@
+              val e = refreshBinders(lambda)(Map.empty, flix)
+              MonoAst.Expr.ApplyClo(e, es, tpe, eff, loc)
+
+            case _ =>
+              val e = visit(exp)
+              MonoAst.Expr.ApplyClo(e, es, tpe, eff, loc)
+          }
+        }
+
+        exp match {
+          case OccurrenceAst1.Expr.Var(sym, _, _) =>
+            varSubst0.get(sym) match {
+              case Some(freshVarSym) => maybeInline(freshVarSym)
+
+              case None =>
+                // If it is the inScopeSet then we have added it via a let-binding so the varSubst should contain sym.
+                // Thus, this is only possible if `sym` is a parameter of the top-level def.
+                val e = visit(exp)
+                MonoAst.Expr.ApplyClo(e, es, tpe, eff, loc)
+            }
+
+          case OccurrenceAst1.Expr.Lambda(fparam, body, _, _) =>
+            // Direct application, e.g., (x -> x)(1)
+            inlineLocalAbstraction(body, List(fparam), es)
+
+          case _ =>
+            val e = visit(exp)
+            MonoAst.Expr.ApplyClo(e, es, tpe, eff, loc)
+        }
+
+      case OccurrenceAst1.Expr.ApplyDef(sym, exps, itpe, tpe, eff, loc) =>
+        val es = exps.map(visit)
+        val def1 = root.defs.apply(sym)
+        // If `def1` is a single non-self call or is trivial
+        // then inline the body of `def1`
+        if (canInlineDef(def1.context, context0)) {
+          inlineDef(def1.exp, def1.fparams, es)
+        } else {
+          MonoAst.Expr.ApplyDef(sym, es, itpe, tpe, eff, loc)
+        }
+
+      case OccurrenceAst1.Expr.ApplyLocalDef(sym, exps, tpe, eff, loc) =>
+        varSubst0.get(sym) match {
+          case Some(freshVarSym) =>
+            val es = exps.map(visit)
+            MonoAst.Expr.ApplyLocalDef(freshVarSym, es, tpe, eff, loc)
+
+          case None => throw InternalCompilerException("unexpected stale local def symbol", loc)
+        }
+
+      case OccurrenceAst1.Expr.Let(sym, exp1, exp2, tpe, eff, occur, loc) =>
+        if (isDead(occur)) {
+          if (isPure(exp1.eff)) {
+            // Case 1:
+            // If `sym` is never used (it is `Dead`)  and `exp1` is pure, so it has no side effects, then it is safe to remove `sym`
+            // Both code size and runtime are reduced
+            visit(exp2)
+          } else {
+            // Case 2:
+            // If `sym` is never used (it is `Dead`) so it is safe to make a Stm.
+            val e1 = visitExp(exp1, varSubst0, subst0, inScopeSet0, context0)
+            MonoAst.Expr.Stm(e1, visit(exp2), tpe, eff, loc)
+          }
+        } else {
+          val freshVarSym = Symbol.freshVarSym(sym)
+          val varSubst1 = varSubst0 + (sym -> freshVarSym)
+
+          // Case 3:
+          // If `exp1` occurs once, and it is pure, then it is safe to inline.
+          // There is a small decrease in code size and runtime.
+          val wantToPreInline = isUsedOnceAndPure(occur, exp1.eff)
```

Comment:
is it on purpose that DontInline is not checked here? i guess it doesn't exist before visiting - is that okay?

Reply:
It implicitly checks `DontInline`